devdoc/job_queue.md

```diff
@@ -0,0 +1,59 @@
+# `job_queue` requirements
+
+`job_queue` is a module that allows jobs to be executed asynchronously. It queues jobs when there are no requests and queues requests when there are no jobs. When a job arrives and there are queued requests, it executes the least recently queued request with the job. When a request arrives and there are queued jobs, it executes the request with the least recently queued job.
+
+Consumers of `job_queue` must create a "`session`" before they can queue requests. A session can make two kinds of requests:
+
+- `get` requests
+- `pop` requests
+
+The consumer provides a `JOB_REQUEST_FUNCTION` that is executed when that request is executed. If the `JOB_REQUEST_FUNCTION` returns `true`, that job is considered completed, and it's  `JOB_COMPLETE_CALLBACK` is called. Once a job is completed, any new sessions that will be started will not see the completed job.
+
+Even after a job is completed, it is not discarded. A job is discarded when all the sessions that were active when the job was pushed have popped it. If all the sessions that were active when the job was queued have popped a job without indicating that the job is complete, the `JOB_COMPLETE_CALLBACK` is called with `JOB_QUEUE_JOB_RESULT_ERROR` as the result.
```

Comment:
>If all the sessions t [](http://example.com/codeflow?start=150&length=21)

Does this mean at least one session must indicate the job is complete?  What if some have popped it without indicating complete, is the result OK or ERROR or something else?

Reply:
At least one session must indicate that the job is complete. If all sessions pop the job without indicating complete, it completed with result ERROR.